### ISSUE TITLE
Will ONNX ever represent a trace of the backprop graph?

### ISSUE THREAD

**Issue by gdiamos** [onnx/onnx#125]
The current implementation seems to only trace forward prop.

**Comment by dzhulgakov** [comment]
We're starting with a clear focus on forward path because training to serving switch provides clear boundary where one might want to switch from one backend to another. It's also the space where hardware vendor libraries usually operate (TensorRT, CoreML, SNPE, etc) and thus there's value in providing a unified interface between them.

Covering the backward path is an eventual goal and we can start scoping out the operators with community at some point.

I want to note that if you're looking for framework-to-framework translation, it's possible to convert the model and use native framework's tools to construct the backward pass.

**Comment by gdiamos** [comment]
This makes sense.  

I'm interested in capturing traces of more of the program so that the traces can be used for full application benchmarking.  

From this perspective it would be valuable to also be able to capture back propagation.

**Comment by binarybana** [comment]
We also would like to see ONNX be flexible enough for this, but rather than adding in new backwards pass ops, I'd propose the approach taken by XLA. In XLA, the [operator set](https://www.tensorflow.org/performance/xla/operation_semantics) is flexible enough to handle both forward and backward passes without needing separate ops* has several advantages:
- The op set that you end up with is more flexible and composable, and isn't restricted to semantics that may not apply in the future.
- The op set is smaller, so less surface area for optimization, backend implementations etc.

`*` - They did for some reason introduce three variants of batchnorm...

**Comment by tjingrant** [comment]
@binarybana it's interesting that you seem to claim XLA does not introduce new backward pass ops. It occurs to me that XLA does exactly that, e.g. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/batch_norm_op.cc . They have registered both `FusedBatchNorm` and `FusedBatchNormGrad` ops. It seems to me that XLA just implicitly creates an *Grad op for every op introduced. 

@binarybana Can you elaborate?

**Comment by binarybana** [comment]
@tjingrant I mentioned in my 'footnote':
> They did for some reason introduce three variants of batchnorm...
But most everything else is represented with first class ops not needing a special purpose operator.

**Comment by tjingrant** [comment]
@binarybana thanks for your clarification. I followed your footnote into batchnorm in the first place but I didn't expect you'd count the corresponding gradient op as merely a `variant`...

**Comment by varunarora** [comment]
Adding my 2 cents (with regards to Baidu's PaddlePaddle).

Firstly, I think ONNX offers an important IR for the case of training too. I have no doubts there; I am not even thinking long term - this is already a need with things like n-Graph. Yes, I understand there is work needed to fully support that. 

I originally was of the favor of having ops for the backward pass within the ONNX graph, but talking more to my team, it doesn't like such a good idea anymore.

Forward pass IRs from frameworks to runtimes should be the only information runtimes actually need. If hardware makers want optimizations for the backward pass, and they are designing backend runtimes to run the forward pass, than it must be onto them on how they build the backward pass from the information they have from the IR's forward pass description.

**Comment by prasanthpul** [comment]
There is now a Training WG that is actively working on a proposal.

**Comment by Linux-cpp-lisp** [comment]
Hi @prasanthpul,

Can you point to more details / are there any updates now almost two years later?

Thanks!

**Comment by gcunhase** [comment]
Any updates as of mid 2022?

**Comment by V0XNIHILI** [comment]
End of 2022 and I have the same question as the original poster. Could someone maybe point to more details/updates to read up on?
